### PR TITLE
Remove alternative main feature

### LINKED ISSUE #29634: Tracking issue for `main` feature
Tracks stabilzation of the `#[main]` attribute, which allows an arbitrary function to be tagged as `main`.

### Status Update

[Decision was reached to **remove** this feature](https://github.com/rust-lang/rust/issues/29634#issuecomment-274425849) and looking for someone to implement; [here is a list of instructions](https://github.com/rust-lang/rust/issues/29634#issuecomment-274951753), feel free to reach out to @nikomatsakis for more tips!

### PR DESCRIPTION
**Pull request by cramertj**
Fix #29634.

WIP because I'm still getting two errors: 
`compile-fail/test-warns-dead-code` is giving an unused attribute error
`compile-fail/issue-12997-2` isn't giving the expected mismatched type error

I'm not sure I understand how either of these are related to the `main` feature, so I'm going to return to this tomorrow after I've gotten some sleep. If anyone has an idea might be causing this, I'd appreciate a hint.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@cramertj hmm the only problem I see on travis is that you have [exceeded the 100 character line limit](https://travis-ci.org/rust-lang/rust/jobs/195802812#L404), which should be relatively easy to fix.

**Comment by cramertj** [comment]
Weird. Must be something messed up locally. Thanks for the heads-up!

**Comment by cramertj** [comment]
@nikomatsakis The issues I mentioned are showing up in travis now that I fixed the line length.

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -468,7 +468,6 @@ pub const BUILTIN_ATTRIBUTES: &'static [(&'static str, AttributeType, AttributeG
 
     ("cfg", Normal, Ungated),
     ("cfg_attr", Normal, Ungated),
-    ("main", Normal, Ungated),
```
Comment (cramertj) [inline src/libsyntax/feature_gate.rs]: Seems weird that this was ungated. Maybe that has to do with the issue.

**Comment by cramertj** [comment]
So, it turns out there's a bunch of stuff internal to the compiler that uses this attribute, so I'm going to close this for now while I look through all of it and figure out what to do. Examples: [here](https://github.com/rust-lang/rust/blob/master/src/libsyntax/test.rs#L485), [here](https://github.com/rust-lang/rust/search?utf8=%E2%9C%93&q=mainattr).

**Pull request closed without merging**